File: cartridges/cartridge_client/src/registry.py
"""Registry client: fetch and parse the app catalog."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cartridge_sdk.net import HttpClient


@dataclass
class RegistryApp:
    """A single app entry from the registry."""

    id: str
    name: str
    description: str = ""
    version: str = "0.1.0"
    author: str = ""
    category: str = "other"
    tags: list[str] = field(default_factory=list)
    repo_url: str = ""
    permissions: list[str] = field(default_factory=list)


@dataclass
class Registry:
    """Parsed registry with category helpers."""

    version: int = 1
    apps: list[RegistryApp] = field(default_factory=list)

    def get_categories(self) -> list[str]:
        cats = sorted({a.category for a in self.apps})
        return cats

    def filter_by_category(self, category: str) -> list[RegistryApp]:
        return [a for a in self.apps if a.category == category]


class RegistryClient:
    """Fetches and parses the registry JSON."""

    def __init__(self, http: HttpClient, url: str) -> None:
        self.http = http
        self.url = url

    async def fetch(self) -> Registry:
        resp = await self.http.get_cached(self.url, ttl_seconds=600)
        if not resp.ok:
            return Registry()

        data = resp.json()
        apps = []
        for entry in data.get("apps", []):
            apps.append(RegistryApp(
                id=entry.get("id", ""),
                name=entry.get("name", ""),
                description=entry.get("description", ""),
                version=entry.get("version", "0.1.0"),
                author=entry.get("author", ""),
                category=entry.get("category", "other"),
                tags=entry.get("tags", []),
                repo_url=entry.get("repo_url", ""),
                permissions=entry.get("permissions", []),
            ))
        return Registry(version=data.get("version", 1), apps=apps)

File: crates/cartridge-net/src/registry.rs
use serde::{Deserialize, Serialize};

use crate::client::HttpClient;

/// A single application entry in the Cartridge registry.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RegistryApp {
    pub id: String,
    pub name: String,
    pub description: String,
    pub version: String,
    pub author: String,
    pub category: String,
    pub tags: Vec<String>,
    pub repo_url: String,
    pub permissions: Vec<String>,
}

/// The full registry payload (matches `registry.json`).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Registry {
    pub version: u32,
    pub apps: Vec<RegistryApp>,
}

impl Registry {
    /// Return a sorted, deduplicated list of all categories present in the
    /// registry.
    pub fn get_categories(&self) -> Vec<String> {
        let mut cats: Vec<String> = self
            .apps
            .iter()
            .map(|a| a.category.clone())
            .collect();
        cats.sort();
        cats.dedup();
        cats
    }

    /// Return all apps that belong to `category`.
    pub fn filter_by_category(&self, category: &str) -> Vec<&RegistryApp> {
        self.apps
            .iter()
            .filter(|a| a.category == category)
            .collect()
    }
}

/// Client for fetching the Cartridge app registry over HTTP.
pub struct RegistryClient {
    http: HttpClient,
    url: String,
}

impl RegistryClient {
    pub fn new(http: HttpClient, url: String) -> Self {
        Self { http, url }
    }

    /// Fetch and parse the registry. Uses a 5-minute cache by default.
    pub fn fetch(&self) -> Result<Registry, String> {
        let response = self.http.get_cached(&self.url, 300)?;
        if !response.ok {
            return Err(format!(
                "Registry request failed with status {}",
                response.status
            ));
        }

        serde_json::from_str::<Registry>(&response.body)
            .map_err(|e| format!("Failed to parse registry JSON: {e}"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_registry_json() {
        let json = r#"{
            "version": 1,
            "apps": [
                {
                    "id": "dev.cartridge.test",
                    "name": "Test App",
                    "description": "A test app",
                    "version": "0.1.0",
                    "author": "Test",
                    "category": "tools",
                    "tags": ["test"],
                    "repo_url": "https://github.com/test/test",
                    "permissions": ["network"]
                },
                {
                    "id": "dev.cartridge.news",
                    "name": "News",
                    "description": "A news app",
                    "version": "0.2.0",
                    "author": "Test",
                    "category": "news",
                    "tags": ["news", "social"],
                    "repo_url": "https://github.com/test/news",
                    "permissions": ["network", "storage"]
                }
            ]
        }"#;

        let registry: Registry = serde_json::from_str(json).unwrap();
        assert_eq!(registry.version, 1);
        assert_eq!(registry.apps.len(), 2);
        assert_eq!(registry.get_categories(), vec!["news", "tools"]);
        assert_eq!(registry.filter_by_category("tools").len(), 1);
        assert_eq!(registry.filter_by_category("tools")[0].id, "dev.cartridge.test");
        assert_eq!(registry.filter_by_category("missing").len(), 0);
    }
}
